File: java/sort/array/QuickSort.java
package sort.array;

import java.util.Random;

public class QuickSort {
    private static final Random _random = new Random();

    public void sort(int[] nums) {
        sort(nums, 0, nums.length - 1);
    }

    private void sort(int[] nums, int start, int end) {
        if (start < end) {
            int partition = getPartition(nums, start, end);
            sort(nums, start, partition - 1);
            sort(nums, partition + 1, end);
        }
    }

    private int getPartition(int[] nums, int start, int end) {
        int index = start + _random.nextInt(end - start + 1);
        swap(nums, index, end);
        int pivot = nums[end];

        int left = start;
        int right = end;
        while (left < right) {
            while (left < right && nums[left] < pivot) {
                ++left;
            }
            while (left < right && nums[right] >= pivot) {
                --right;
            }

            if (left < right) {
                swap(nums, left, right);
            }
        }
        swap(nums, left, end);
        return left;
    }

    private void swap(int[] nums, int i, int j) {
        if (i != j) {
            nums[i] ^= nums[j];
            nums[j] ^= nums[i];
            nums[i] ^= nums[j];
        }
    }
}

File: py/sort/array/QuickSort.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import random
from typing import List

class QuickSort:
    def sort(self, nums: List[int]) -> None:
        self.__sort(nums, 0, len(nums) - 1)

    def __sort(self, nums: List[int], start: int, end: int) -> None:
        if start < end:
            partition = self.__getPartition(nums, start, end)
            self.__sort(nums, start, partition - 1)
            self.__sort(nums, partition + 1, end)

    def __getPartition(self, nums: List[int], start: int, end: int) -> int:
        index = random.randrange(start, end + 1)
        nums[index], nums[end] = nums[end], nums[index]
        pivot = nums[end]

        left, right = start, end
        while left < right:
            while left < right and nums[left] < pivot:
                left += 1
            while left < right and nums[right] >= pivot:
                right -= 1

            if left < right:
                nums[left], nums[right] = nums[right], nums[left]
        nums[left], nums[end] = nums[end], nums[left]
        return left
